[WIP] Multithreading: Threads, Synchronization, Atomics and Futures

valdisxp1 opened the pull request:
Includes #1042 and #1035.

# What's new:
## Functionality
1. Thread creation and `Thread.join`
2.  `synchronized` blocks, `wait`-`notify`
3. `Thread.sleep` with `Thread.interrupt` support
4. `ThreadLocal` 
5. fixed `java.lang.AtomicReference` and `java.lang.AtomicReferenceArray`
6. fixed state in `AbstractPromise` by making it thread-safe
7. global `ExecutionContext` uses all available cores and `Future` s are running in parallel with the main Thread
8. made `StackTraceElement` thread-safe
9. added missing `System.setIn`, `System.setOut`, `System.setErr` methods needed for testing 
10. bound more methods and constants from `sysinfo`, `pthread`, `time` and `signal`
11. A Scala-native program now automatically shuts down when there are no non-daemon threads running, just like Scala/Java on JVM. For comparison, a C/C++ program on posix will automatically shutdown when main thread ends.

See full details in the attached Gist below.
## GC
1. made "none" GC Thread-safe (and wait-free)
2. configured "boehm" GC for multithreading
3.  **"immix" GC does not support multithreading yet.** Attempting to create a thread under immix will stop the program with an error. For this reason "immix" is now excluded from running automated tests.
## Testing
1. Added `MultiThreadSuite` trait with utilities for multi-threaded testing
  a. simple utilities: `takesAtLeast(ms,ns)(function)`,  `withThreads(numTreads)(function)` and `withExceptionHandler`
  b. testing constructions `eventually`, `eventuallyEquals`, `eventuallyConstant`
  c. self-adjusting constructions `testWithMinDelay` and `testWithMinRepetitions`
  d. useful threads`WaitingThread(mutex)`, stoppable `Counter` and `MemoryMuncher`
2. Almost every functionality has an automated test. See full details in the attached Gist below.
3. Atomics are also automatically tested for atomicity by running a lot of operations concurrently.

# Known issues:
1. Uncaught exception in `synchronized` block, which is caught outside it, does not immediately release the lock. Note that when the thread terminates, it tries to release all the locks it might be holding, which prevents issues in some cases.
   a. **Workaround**: the exception can be passed "around" the  `synchronized` See  `ShadowLock` in `Monitor.scala` for an example.
2. Boehm might break memory consistency for some multi-threaded programs. For example, something Future-heavy like one of my benchmarks.
> Thread users should also be aware that on many platforms objects reachable only from thread-local variables may be prematurely reclaimed. Thus objects pointed to by thread-local variables should also be pointed to by a globally visible data structure. (This is viewed as a bug, but as one that is exceedingly hard to fix without some libc hooks.)
http://www.hboehm.info/gc/gcinterface.html
3. `Thread.suspend` can fail and getting the stack of another thread may block if the `signal` fails to be delivered. In **some environments** even real-time signals might **sometimes** fail to deliver. This is the case with the Travis-CI server. 
  a. **Workaround**: Repeat `Thread.suspend` until you get the desired result. For getting stacktraces: use two threads: one for reading the results (this one may block) and another one to resend the signal manually if there are no results.
4. `Thread.join()` (the one without the timeout) might block indefinitely even when that thread has finished. It happened 1 in 1000 times of running a suit with every multi-threaded test (extremely rare). The underlying issue might even be fixed as I have not seen it in a while.
  a. **Workaround**:  Use `Thread.join` with timeout. Wrap in a while loop if necessary.
5. Memory/resource leak when using `synchronized`. Mutexes and cond objects are not destroyed when the object on which you synchronize is garbage collected. Currently, there is no finalizers or special GC treatment for Monitor objects.

Detailed progress: https://gist.github.com/valdisxp1/b0127ee52e90b42835dbdabbf6a31d04

valdisxp1 commented:
Note, that the tests are failing because of immix

valdisxp1 commented:
The only test that is failing is `"*DEPRECATED*  ThreadGroup.suspend and resume should affect all threads"` in `ThreadGroupSuite`. However, the other `suspend` tests pass. Therefore, `ThreadGroup.suspend` may not suspend the thread in some cases.

The issue is the signal is sent, but not always received. This is why real-time signals (see signal(7)) are now used. They better delivery guarantees, like preserving the order. This fixed the issue for my local machine. However, it is possible real-time signals are implemented differently in the kernel/virtualization environment of Travis. Or they may be not implemented at all.

tindzk commented on tools/src/main/scala/scala/scalanative/optimizer/pass/StringLowering.scala:
```diff
@@ -18,7 +18,7 @@ class StringLowering(implicit top: Top) extends Pass {
     val node  = ClassRef.unapply(StringName).get
     val names = node.layout.entries.map(_.name)
-    assert(names.length == 4, "java.lang.String is expected to have 4 fields")
+    assert(names.length == 5, "java.lang.String is expected to have 4 fields")
```
The message should be updated too: "[...] **5** fields"

valdisxp1 replied:
@tindzk Nice catch! I fixed it in the latest.

valdisxp1 commented:
@shadaj thanks for your feedback. It is such a cool feeling to see someone already using this :smile: 
About casting the Thread to a pointer. Nice catch! :+1:  I will add this change asap.
I'll have too look in more detail about your other comments.

valdisxp1 commented:
- @shadaj I removed the recursive `Atomic.free()` and wired the memory operations directly to `stdlib`. This should resolve the issue with `-O0`.
  - about `AbstractQueuedSynchronizer` breaking. It is expected that everything in `java.util.concurrent` apart from the atomics does not work properly. It has not been tested.
  - Again thanks for the idea. I implemented the direct casts from the `this` reference. While doing that I found a few issues new issues that needed to be fixed.
- Monitor methods are working as expected - same as in Java. 
> Throws:
> IllegalMonitorStateException - if the current thread is not the owner of the object's monitor.
> https://docs.oracle.com/javase/7/docs/api/java/lang/Object.html#wait() 
Scala on JVM also acts the same both on amd64 and arm.
- I am not sure about inlining the implicit conventions. In my honest opinion, the java atomics code will become harder to read. It might be better have the additional changes to Atomics as a separate PR. There might be quite a lot besides inlining the implicits.

ekrich commented:
@valdisxp1  Did you mean `overflowNanos`? https://github.com/valdisxp1/scala-native/blob/e77e39e7b34bc6547a8bbf8932dcd6286d04c2da/nativelib/src/main/scala/scala/scalanative/runtime/Monitor.scala#L57

valdisxp1 commented:
@ekrich I intentionally named it  `overflown`, but it makes sense as `overflow` as well.

ekrich commented:
@LukasKellenberger It looks like we are getting closer to supporting threads - Do you know what it would take for Immix to support threading? I think Immix is slated to be default in the next release.

LukasKellenberger commented:
@ekrich Allocation shouldn't be too hard, we need to have a global allocator that handles allocation on a block level that would be synchronised and then one local allocator per thread that doesn't need to be synchronized. For the heap of large objects, I guess we could have it fully synchronised.

Collection is another story. For that I think that we would need safe points (As far as I remember @densh has a branch with safe points). These safe points are needed to tell all threads to stop working and pause until collection is done.

ekrich commented:
@LukasKellenberger Thanks for the update and your work on Immix 😄

lolgab commented:
> Collection is another story. For that I think that we would need safe points (As far as I remember @densh has a branch with safe points). These safe points are needed to tell all threads to stop working and pause until collection is done.

Hi @LukasKellenberger, 
For what I understood, we don't necessarily need safe points to make immix working with multiple threads. Since Immix stack scanning is conservative (not sure about this assumption) we could mimic what boehm does to stop the world without needing changes in the code generation.
It uses async signal passing between threads using `pthread_kill` to tell them to stop, gets signaled when all of them are stopped, then scan all stacks got using `pthread_attr_getstack`.
Although it is harder than what it seems, it's not impossible to modify immix to work with threads maintaining the current structure and stopping the world using signals.

lolgab commented:
I added a pull request that makes Immix thread aware, allocating in mutex block and stopping all threads during collection (using `pthread_kill` and signal handlers).
You can find the PR [here](https://github.com/scala-native/scala-native/pull/1726/files).
I would be happy to receive suggestions and comments :)

LeeTibbert commented:
@valdisxp1 @Korf74  May I have your permission to use the atomic related files of this PR as a baseline
for work to add atomics?   Doing so is likely to save me weeks of time and make adding 
atomics doable with the time & energy I have available.

My driving force is that, due to demands a the time, I have some code in javalib
`concurrent.atomic.*.scala` which is clearly documented as non-atomic but
which is ~~a hazard~~ technical debt lying around with my fingerprints
on it. 

 I am particularly interested in evolving the .gyb files and studying the unit-tests.
It looks to me like there is a lot of good, detailed, hard work in this PR.  My dream is to evolve
it and submit some descendant of it.  

You may or may not recognize anything I submit, but I will know, and will acknowledge,
the origins.

 I may be asked/directed to do a 'light' version, in `clib` rather than supporting CAtomics
as a first class citizen.  I thought I would ask your permission to evolve your intellectual property
before kicking off that design discussion with the current community & Scala Center leadership.  

Thank you for your time.
